runtime/lua/vim/lsp/structures/diagnostic.lua

```diff
@@ -0,0 +1,230 @@
+local api = vim.api
+local validate = vim.validate
+
+local highlight = require('vim.highlight')
+local protocol = require('vim.lsp.protocol')
+
+local Position = require('vim.lsp.structures.position')
+
+local underline_highlight_name = "LspDiagnosticsUnderline"
+
+local hlmap = {
+  [protocol.DiagnosticSeverity.Error]       = underline_highlight_name .. 'Error',
+  [protocol.DiagnosticSeverity.Warning]     = underline_highlight_name .. 'Warning',
+  [protocol.DiagnosticSeverity.Information] = underline_highlight_name .. 'Information',
+  [protocol.DiagnosticSeverity.Hint]        = underline_highlight_name .. 'Hint',
+}
+
+local diagnostic_severity_map = {
+  [protocol.DiagnosticSeverity.Error] = "LspDiagnosticsErrorSign";
+  [protocol.DiagnosticSeverity.Warning] = "LspDiagnosticsWarningSign";
+  [protocol.DiagnosticSeverity.Information] = "LspDiagnosticsInformationSign";
+  [protocol.DiagnosticSeverity.Hint] = "LspDiagnosticsHintSign";
+}
+
+local default_severity_highlight = {
+  [protocol.DiagnosticSeverity.Error] = { guifg = "Red" };
+  [protocol.DiagnosticSeverity.Warning] = { guifg = "Orange" };
+  [protocol.DiagnosticSeverity.Information] = { guifg = "LightBlue" };
+  [protocol.DiagnosticSeverity.Hint] = { guifg = "LightGrey" };
+}
+
+local severity_highlights = {}
+
+-- Initialize default severity highlights
+for severity, hi_info in pairs(default_severity_highlight) do
+  local severity_name = protocol.DiagnosticSeverity[severity]
+  local highlight_name = "LspDiagnostics"..severity_name
+  -- Try to fill in the foreground color with a sane default.
+  local cmd_parts = {"highlight", "default", highlight_name}
+  for k, v in pairs(hi_info) do
+    table.insert(cmd_parts, k.."="..v)
+  end
+  api.nvim_command(table.concat(cmd_parts, ' '))
+  api.nvim_command('highlight link ' .. highlight_name .. 'Sign ' .. highlight_name)
+  severity_highlights[severity] = highlight_name
+end
```

Comment:
Basically all the stuff above could probably use the config table approach to override if necessary (or maybe even add per-server configuration? We don't have that yet though)